tests/test_compare_dictconfig_vs_dict.py

```diff
@@ -0,0 +1,544 @@
+        False,
+        Enum1.FOO,
+        Enum1.BAR,
+        "FOO",
+        "BAR",
+    ]
+)
+def key(request: Any) -> Any:
+    """A key to test indexing into DictConfig."""
+    return request.param
+
+
+@pytest.fixture
+def python_dict(data: Dict[Any, Any]) -> Dict[Any, Any]:
+    """Just a standard python dictionary, to be used in comparison with DictConfig."""
+    return deepcopy(data)
+
+
+@pytest.fixture(params=[True, False])
+def struct_mode(request: Any) -> bool:
+    struct_mode: bool = request.param
+    return struct_mode
+
+
+@pytest.mark.parametrize(
+    "data",
+    [
+        pytest.param({"a": 10}, id="str"),
+        pytest.param({1: "a"}, id="int"),
+        pytest.param({123.45: "a"}, id="float"),
+        pytest.param({True: "a"}, id="bool"),
+        pytest.param({Enum1.FOO: "foo"}, id="Enum1"),
+    ],
+)
+class TestUntypedDictConfig:
+    """Compare DictConfig with python dict in the case where key_type is not set."""
+
+    @pytest.fixture
+    def cfg(self, python_dict: Any, struct_mode: bool) -> DictConfig:
+        """Create a DictConfig instance from the given data"""
+        cfg: DictConfig = DictConfig(content=python_dict)
+        if struct_mode:
+            OmegaConf.set_struct(cfg, True)
+        return cfg
+
+    def test__setitem__(
+        self, python_dict: Any, cfg: DictConfig, key: Any, struct_mode: bool
+    ) -> None:
+        """Ensure that __setitem__ has same effect on python dict and on DictConfig."""
+        if struct_mode and key not in cfg:
+            with pytest.raises(ConfigKeyError):
+                cfg[key] = "sentinel"
+        else:
+            python_dict[key] = "sentinel"
+            cfg[key] = "sentinel"
+            assert python_dict == cfg
+
+    def test__getitem__(self, python_dict: Any, cfg: DictConfig, key: Any) -> None:
+        """Ensure that __getitem__ has same result with python dict as with DictConfig."""
+        try:
+            result = python_dict[key]
+        except KeyError:
+            with pytest.raises(ConfigKeyError):
+                cfg[key]
+        else:
+            assert result == cfg[key]
+
+    def test__delitem__(
+        self, python_dict: Any, cfg: DictConfig, key: Any, struct_mode: bool
+    ) -> None:
+        """Ensure that __delitem__ has same result with python dict as with DictConfig."""
+        if struct_mode:
+            with pytest.raises(ConfigTypeError):
+                del cfg[key]
+        else:
+            try:
+                del python_dict[key]
+                assert key not in python_dict
+            except KeyError:
+                with pytest.raises(ConfigKeyError):
+                    del cfg[key]
+            else:
+                del cfg[key]
+                assert key not in cfg
+
+    def test__contains__(self, python_dict: Any, cfg: Any, key: Any) -> None:
+        """Ensure that __contains__ has same result with python dict as with DictConfig."""
+        assert (key in python_dict) == (key in cfg)
```

Comment:
This test file has a very large surface area. It could be a good idea to followup with a second cleanup pass in a different PR.

Reply:
> I think this approach scales better if we can pull it off.

I think that pulling this off will be a good exercise for me to learn more about `pytest`.

Reply:
I spent 15 minutes now to see if I can get something decent.
this is what I came up with:
```python
from typing import MutableMapping, Any, List
from pytest import raises, mark

from omegaconf import OmegaConf, DictConfig


def struct(proto: Any) -> DictConfig:
    cfg = OmegaConf.create(proto)
    OmegaConf.set_struct(cfg, True)
    return cfg


def dictconfig(proto: Any) -> DictConfig:
    return OmegaConf.create(proto)


def dicts(proto: Any, factories: List[Any]) -> Any:
    return [f(proto) for f in factories]


class TestDict:
    @mark.parametrize("cfg", dicts({}, [dict, OmegaConf.create]))
    def test_pop_empty(self, cfg: MutableMapping) -> None:
        with raises(KeyError):
            cfg.pop("not_found")

    @mark.parametrize("cfg", dicts({"foo": 10}, [dict, dictconfig]))
    def test_pop(self, cfg: MutableMapping):
        assert cfg.pop("foo") == 10

    @mark.parametrize("cfg", dicts({"foo": 10}, [dict, dictconfig]))
    def test_pop_with_default(self, cfg: MutableMapping):
        assert cfg.pop("foo", 20) == 10

    @mark.parametrize("cfg", dicts({"foo": 10}, [dict, dictconfig]))
    def test_pop_with_default2(self, cfg: MutableMapping):
        assert cfg.pop("not_found", 10) == 10
        assert cfg == {"foo": 10}

    @mark.parametrize("cfg", dicts({"foo": 10}, [dict, dictconfig, struct]))
    def test_getitem(self, cfg: MutableMapping):
        assert cfg["foo"] == 10
```

Reply:
Hmm...

I'm not sure if it will be so easy to convert the existing test file into this format.
In the current setup, a large number of test cases are automatically generated in `test_compare_dictconfig_vs_dict.py`.
For example, many of the tests in the `TestUntypedDictConfig` class are run once with each possible combination of the following inputs:
```
for key in ["a", "b", 1, 2, 123.45, 67.89, True, False, Enum1.FOO, Enum1.BAR, "FOO", "BAR"]:
    for data in [{"a": 10}, {1: "a"}, {123.45: "a"}, {True: "a"}, {Enum1.FOO: "foo"}]:
        for struct_mode in [True, False]:
            ...
```

Thus, most of the methods in the `TestUntypedDictConfig` class gets called 120 times:
12 keys x 5 data x 2 struct_mode. (A few of the methods get called only 10 times each because they don't depend on the `key` fixture). So in total, pytest runs 1110 tests for the `TestUntypedDictConfig` class. And then there are the other classes, giving a total of 2220 tests in the whole `test_compare_dictconfig_vs_dict.py` file. This number will increase once more of the tests have been split to handle separate cases in atomic tests.

My feeling is that this large amount of tests is made possible by making the tests depend on both the `DictConfig` object and on the standard python dict at the same time. This way, the programmer does not have to write down what is expected in every test case. Instead, the python dict is used to figure out what is expected, and the programmer just has to define how the particular (key, python_dict) combination should be used to figure out what is expected of the `DictConfig` object.

Reply:
Fair enough :P

Reply:
> Tests that are smart are easier to write but harder to maintain and understand.
> We are only writing them once, but we will be maintaining them forever.
> They also provide better documentation to the expected behavior.

Good points.

Reply:
> Also, not all tests behaves the same with and without struct mode (which can be expressed in my version by opting out of testing struct, and adding an explicit test to cover it).

I found a cool way to opt out of specific cases using the "smart" tests: you write a general `struct_mode` fixture, then override it for specific tests:
```python3
from pytest import fixture, mark
from omegaconf import OmegaConf

data_dict_examples = [{"a": 123}]

@fixture(params=[True, False])
def struct_mode(request) -> bool:
    struct_mode: bool = request.param
    return struct_mode

@fixture(params=data_dict_examples)
def dictconfig(request, struct_mode) -> bool:
    cfg = OmegaConf.create(request.param)
    if struct_mode:
        OmegaConf.set_struct(cfg, True)
    return cfg

...  # here are lots of tests that use dictconfig with and without struct_mode

# Here we override struct_mode for specific tests:

@mark.parametrize("struct_mode", [False])
def test_for_specific_non_struct_behavior(dictconfig):
    ...  # Here dictconfig is constructed with struct_mode=False only

@mark.parametrize("struct_mode", [True])
def test_for_specific_non_struct_behavior(dictconfig):
    ...  # Here dictconfig is constructed with struct_mode=True only
```
Anyway, if we're going to switch to a simpler testing strategy then this isn't totally relevant. I just wanted to share the pytest trick :)

Reply:
Very nice.

By the way:
flags has 3 states:
None, True and False.